File: csharp/src/Vista.SDK/Gmod.Traversal.cs
namespace Vista.SDK;

public delegate TraversalHandlerResult TraverseHandler(IReadOnlyList<GmodNode> parents, GmodNode node);
public delegate TraversalHandlerResult TraverseHandlerWithState<TState>(
    TState state,
    IReadOnlyList<GmodNode> parents,
    GmodNode node
);

public enum TraversalHandlerResult
{
    Stop,
    SkipSubtree,
    Continue,
}

public record TraversalOptions()
{
    internal const int DEFAULT_MAX_TRAVERSAL_OCCURRENCE = 1;

    /// <summary>The maximum number of times a node can occur in a path. The traversal will stop and the first node to reach the limit, and include it as the end node. Increasing this will drastically reduce performance</summary>
    /// <remarks>The default value is 1.</remarks>
    /// <example>411.1/C101.63/S206.22/S110.2/C101 = MaxTraversalOccurrence 1. The travesal algorithm finds the second instance of C101 in the path, and stop, but include the node in the result.</example>
    public int MaxTraversalOccurrence { get; set; } = DEFAULT_MAX_TRAVERSAL_OCCURRENCE;
}

public sealed partial class Gmod
{
    public bool Traverse(TraverseHandler handler, TraversalOptions? options = null) =>
        Traverse(handler, _rootNode, (handler, parents, node) => handler(parents, node), options);

    public bool Traverse<TState>(
        TState state,
        TraverseHandlerWithState<TState> handler,
        TraversalOptions? options = null
    ) => Traverse(state, _rootNode, handler, options);

    public bool Traverse(GmodNode rootNode, TraverseHandler handler, TraversalOptions? options = null) =>
        Traverse(handler, rootNode, (handler, parents, node) => handler(parents, node), options);

    public bool Traverse<TState>(
        TState state,
        GmodNode rootNode,
        TraverseHandlerWithState<TState> handler,
        TraversalOptions? options = null
    )
    {
        var opts = options ?? new TraversalOptions();
        var context = new TraversalContext<TState>(new Parents(), handler, state, opts.MaxTraversalOccurrence);
        return TraverseNode(in context, rootNode) == TraversalHandlerResult.Continue;
    }

    private TraversalHandlerResult TraverseNode<TState>(in TraversalContext<TState> context, GmodNode node)
    {
        if (node.Metadata.InstallSubstructure == false)
            return TraversalHandlerResult.Continue;

        var result = context.Handler(context.State, context.Parents.AsList, node);
        if (result is TraversalHandlerResult.Stop or TraversalHandlerResult.SkipSubtree)
            return result;

        var skipOccurenceCheck = IsProductSelectionAssignment(context.Parents.LastOrDefault(), node);
        // Skip the occurence check for "hidden" nodes such as selections, etc.
        if (!skipOccurenceCheck)
        {
            var occ = context.Parents.Occurrences(node);
            if (occ == context.MaxTraversalOccurrence)
                return TraversalHandlerResult.SkipSubtree;
            if (occ > context.MaxTraversalOccurrence)
                throw new Exception("Invalid state - node occured more than expected");
        }
        context.Parents.Push(node);

        for (int i = 0; i < node.Children.Count; i++)
        {
            var child = node.Children[i];
            result = TraverseNode(in context, child);
            if (result is TraversalHandlerResult.Stop)
                return result;
            else if (result is TraversalHandlerResult.SkipSubtree)
                continue;
        }

        context.Parents.Pop();
        return TraversalHandlerResult.Continue;
    }

    internal bool PathExistsBetween(
        IEnumerable<GmodNode> fromPath,
        GmodNode to,
        out IEnumerable<GmodNode> remainingParents
    )
    {
        var lastAssetFunction = fromPath.LastOrDefault(n => n.IsAssetFunctionNode);
        remainingParents =  [];

        var state = new PathExistsContext(to) { RemainingParents = remainingParents, FromPath =  [.. fromPath] };

        var reachedEnd = Traverse(
            state,
            lastAssetFunction ?? RootNode,
            (state, parents, node) =>
            {
                if (node.Code != state.To.Code)
                    return TraversalHandlerResult.Continue;
                List<GmodNode>? actualParents = null;
                while (!parents[0].IsRoot)
                {
                    if (actualParents is null)
                    {
                        actualParents =  [.. parents];
                        parents = actualParents;
                    }

                    var parent = parents[0];
                    if (parent.Parents.Count != 1)
                        throw new Exception("Invalid state - expected one parent");

                    actualParents.Insert(0, parent.Parents[0]);
                }

                // Validate parents
                if (parents.Count < state.FromPath.Count)
                    return TraversalHandlerResult.Continue;
                // Must have same start order
                var match = true;
                for (int i = 0; i < state.FromPath.Count; i++)
                {
                    if (parents[i].Code != state.FromPath[i].Code)
                    {
                        match = false;
                        break;
                    }
                }

                if (match)
                {
                    state.RemainingParents = parents.Where(p => !state.FromPath.Any(pp => pp.Code == p.Code)).ToArray();
                    return TraversalHandlerResult.Stop;
                }

                return TraversalHandlerResult.Continue;
            }
        );

        remainingParents = state.RemainingParents;

        return !reachedEnd;
    }

    record PathExistsContext(GmodNode To)
    {
        public IEnumerable<GmodNode> RemainingParents = Enumerable.Empty<GmodNode>();
        public required IReadOnlyList<GmodNode> FromPath { get; init; }
    }

    private readonly record struct TraversalContext<TState>(
        Parents Parents,
        TraverseHandlerWithState<TState> Handler,
        TState State,
        int MaxTraversalOccurrence
    );

    private readonly struct Parents
    {
        private readonly Dictionary<string, int> _occurrences = new(4);
        private readonly List<GmodNode> _parents = new(64);

        public Parents() { }

        public readonly void Push(GmodNode parent)
        {
            _parents.Add(parent);
            if (_occurrences.ContainsKey(parent.Code))
                _occurrences[parent.Code]++;
            else
                _occurrences.Add(parent.Code, 1);
        }

        public readonly void Pop()
        {
            var parent = _parents[_parents.Count - 1];
            _parents.RemoveAt(_parents.Count - 1);

            if (_occurrences.TryGetValue(parent.Code, out var occ))
            {
                if (occ == 1)
                    _occurrences.Remove(parent.Code);
                else
                    _occurrences[parent.Code]--;
            }
        }

        public readonly int Occurrences(GmodNode node) => _occurrences.TryGetValue(node.Code, out var occ) ? occ : 0;

        public readonly GmodNode? LastOrDefault() => _parents.Count > 0 ? _parents[_parents.Count - 1] : null;

        public readonly IReadOnlyList<GmodNode> ToList() => _parents.ToList();

        public readonly IReadOnlyList<GmodNode> AsList => _parents;
    }
}

File: cpp/src/dnv/vista/sdk/GmodTraversal.cpp
#include "pch.h"

#include "dnv/vista/sdk/GmodTraversal.h"

#include "dnv/vista/sdk/utils/StringUtils.h"

namespace dnv::vista::sdk
{
	//=====================================================================
	// Traversal algorithms
	//=====================================================================

	namespace GmodTraversal
	{
		namespace detail
		{
			//----------------------------------------------
			// Path analysis context
			//----------------------------------------------

			struct PathExistsContext
			{
				const GmodNode& to;
				std::vector<const GmodNode*> remainingParents;
				std::vector<const GmodNode*> fromPath;

				PathExistsContext( const GmodNode& toNode, const std::vector<const GmodNode*>& fromPathList )
					: to{ toNode }, fromPath{ fromPathList }
				{
					remainingParents.reserve( 32 );
				}

				PathExistsContext() = delete;
				PathExistsContext( const PathExistsContext& ) = delete;
				PathExistsContext( PathExistsContext&& ) noexcept = delete;
				~PathExistsContext() = default;
				PathExistsContext& operator=( const PathExistsContext& ) = delete;
				PathExistsContext& operator=( PathExistsContext&& ) noexcept = delete;
			};
		}

		//=====================================================================
		// Public API
		//=====================================================================

		bool traverse( const Gmod& gmodInstance, TraverseHandler handler, const TraversalOptions& options )
		{
			TraverseHandler capturedHandler = handler;
			TraverseHandlerWithState<TraverseHandler> wrapperHandler =
				[]( TraverseHandler& h, const std::vector<const GmodNode*>& parents, const GmodNode& node )
				-> TraversalHandlerResult { return h( parents, node ); };

			return traverse( capturedHandler, gmodInstance, wrapperHandler, options );
		}

		bool traverse( const GmodNode& rootNode, TraverseHandler handler, const TraversalOptions& options )
		{
			TraverseHandler capturedHandler = handler;
			TraverseHandlerWithState<TraverseHandler> wrapperHandler =
				[]( TraverseHandler& h, const std::vector<const GmodNode*>& parents, const GmodNode& node )
				-> TraversalHandlerResult { return h( parents, node ); };

			return traverse( capturedHandler, rootNode, wrapperHandler, options );
		}

		bool pathExistsBetween(
			const Gmod& gmodInstance,
			const std::vector<const GmodNode*>& fromPath,
			const GmodNode& to,
			std::vector<const GmodNode*>& remainingParents )
		{
			remainingParents.clear();

			const GmodNode* lastAssetFunction = nullptr;
			for ( auto it = fromPath.rbegin(); it != fromPath.rend(); ++it )
			{
				if ( ( *it )->metadata().category() == gmod::GMODNODE_CATEGORY_ASSET_FUNCTION )
				{
					lastAssetFunction = *it;
					break;
				}
			}

			detail::PathExistsContext context( to, fromPath );
			TraverseHandlerWithState<detail::PathExistsContext> handler =
				[]( detail::PathExistsContext& state, const std::vector<const GmodNode*>& parents, const GmodNode& node ) -> TraversalHandlerResult {
				if ( std::string_view( node.code() ) != state.to.code() )
				{
					return TraversalHandlerResult::Continue;
				}

				std::vector<const GmodNode*> actualParents;
				actualParents.reserve( parents.size() + 16 );

				if ( !parents.empty() && !parents[0]->isRoot() )
				{
					actualParents.assign( parents.begin(), parents.end() );

					while ( !actualParents.empty() && !actualParents[0]->isRoot() )
					{
						const GmodNode* parent = actualParents[0];
						const auto& parentOfCurrent = parent->parents();
						if ( parentOfCurrent.size() != 1 )
						{
							break;
						}
						actualParents.insert( actualParents.begin(), parentOfCurrent[0] );
					}
				}
				else
				{
					actualParents.assign( parents.begin(), parents.end() );
				}

				if ( actualParents.size() < state.fromPath.size() )
				{
					return TraversalHandlerResult::Continue;
				}

				bool match = true;
				for ( size_t i = 0; i < state.fromPath.size(); ++i )
				{
					if ( actualParents[i]->code() != state.fromPath[i]->code() )
					{
						match = false;
						break;
					}
				}

				if ( match )
				{
					state.remainingParents.clear();
					state.remainingParents.reserve( actualParents.size() );

					for ( const GmodNode* parent : actualParents )
					{
						bool existsInFromPath = false;
						for ( const GmodNode* fromPathNode : state.fromPath )
						{
							if ( parent->code() == fromPathNode->code() )
							{
								existsInFromPath = true;
								break;
							}
						}
						if ( !existsInFromPath )
						{
							state.remainingParents.push_back( parent );
						}
					}
					return TraversalHandlerResult::Stop;
				}

				return TraversalHandlerResult::Continue;
			};
			bool reachedEnd = traverse(
				context,
				lastAssetFunction ? *lastAssetFunction : gmodInstance.rootNode(),
				handler );

			remainingParents = std::move( context.remainingParents );

			return !reachedEnd;
		}
	}
}
